sdk/azidentity/chained_token_credential.go

```diff
@@ -50,35 +54,72 @@ func NewChainedTokenCredential(sources []azcore.TokenCredential, options *Chaine
 	if options == nil {
 		options = &ChainedTokenCredentialOptions{}
 	}
-	return &ChainedTokenCredential{sources: cp, name: "ChainedTokenCredential", retrySources: options.RetrySources}, nil
+	mut := sync.RWMutex{}
+	return &ChainedTokenCredential{
+		cond:         sync.NewCond(&mut),
+		mut:          &mut,
+		sources:      cp,
+		name:         "ChainedTokenCredential",
+		retrySources: options.RetrySources,
+	}, nil
 }
 
 // GetToken calls GetToken on the chained credentials in turn, stopping when one returns a token. This method is called automatically by Azure SDK clients.
 // ctx: Context controlling the request lifetime.
 // opts: Options for the token request, in particular the desired scope of the access token.
 func (c *ChainedTokenCredential) GetToken(ctx context.Context, opts policy.TokenRequestOptions) (*azcore.AccessToken, error) {
-	if c.successfulCredential != nil && !c.retrySources {
-		return c.successfulCredential.GetToken(ctx, opts)
+	if !c.retrySources {
+		// ensure only one goroutine at a time iterates the sources and perhaps sets c.successfulCredential
+		for {
+			c.mut.RLock()
+			mustIterate := c.successfulCredential == nil
+			c.mut.RUnlock()
+			if !mustIterate {
+				return c.successfulCredential.GetToken(ctx, opts)
+			}
+			c.cond.L.Lock()
+			defer c.cond.L.Unlock()
```

Comment:
I don't think this is correct (it's safe but doesn't take advantage of the condition).  The exclusive lock should be released once the for loop exits so that other goroutines will hit `c.cond.Wait()`.

Reply:
You're right, the condition isn't effective here. I'll have to give it more thought. At the moment I'm thinking my read locking approach prevents straightforward use of `Cond` because there are really two conditions here, only one of which is checked under an exclusive lock as `Cond` expects. If that's the case, I would rather give up on `Cond` than read locking.